pkg/sources/udsource/user_defined_source.go

```diff
@@ -130,15 +121,28 @@ func (u *userDefinedSource) GetName() string {
 	return u.vertexName
 }
 
-// GetPartitionIdx returns the partition number for the user-defined source.
-// Source is like a buffer with only one partition. So, we always return 0
-func (u *userDefinedSource) GetPartitionIdx() int32 {
-	return 0
+// Partitions returns the partitions of the user-defined source
+func (u *userDefinedSource) Partitions() []int32 {
+	partitions := make([]int32, 0, len(u.partitions))
+	for partition := range u.partitions {
+		partitions = append(partitions, partition)
+	}
+	return partitions
 }
 
-// Read reads the messages from the user-defined source
+// Read reads the messages from the user-defined source, tracks the partitions from which the messages are read
+// tracked partitions are used to determine the partitions to which the watermarks should be published
 func (u *userDefinedSource) Read(ctx context.Context, count int64) ([]*isb.ReadMessage, error) {
-	return u.sourceApplier.ApplyReadFn(ctx, count, u.readTimeout)
+	readMessages, err := u.sourceApplier.ApplyReadFn(ctx, count, u.readTimeout)
+	if err != nil {
+		return nil, err
+	}
+	for _, msg := range readMessages {
+		if _, ok := u.partitions[msg.ReadOffset.PartitionIdx()]; !ok {
+			u.partitions[msg.ReadOffset.PartitionIdx()] = true
+		}
```

Comment:
nit: would it be simpler if we 
```suggestion
		u.partitions[msg.ReadOffset.PartitionIdx()] = true
```

Reply:
why perform a write/update operation if the key is already present?

Reply:
```
func BenchmarkMapUpdate(b *testing.B) {
	tempMap := make(map[string]struct{})

	for i := 0; i < b.N; i++ {
		tempMap["test"] = struct{}{}
	}
	b.ReportAllocs()
}

func BenchmarkMapReadAndUpdate(b *testing.B) {
	tempMap := make(map[string]struct{})

	for i := 0; i < b.N; i++ {
		if _, ok := tempMap["test"]; !ok {
			tempMap["test"] = struct{}{}
		}
	}
	b.ReportAllocs()
}

//goos: darwin
//goarch: arm64
//pkg: github.com/numaproj/numaflow/pkg/window/strategy/sliding
//BenchmarkMapUpdate
//BenchmarkMapUpdate-10    	154240488	         7.747 ns/op	       0 B/op	       0 allocs/op
//PASS

//goos: darwin
//goarch: arm64
//pkg: github.com/numaproj/numaflow/pkg/window/strategy/sliding
//BenchmarkMapReadAndUpdate
//BenchmarkMapReadAndUpdate-10    	442961821	         2.655 ns/op	       0 B/op	       0 allocs/op
//PASS
```

Reply:
Thanks for running the benchmark. This proves that `checking before adding` have a performance advantage in the case of a large map with many key collisions. I still think for our case(small map with many key collision), `directly adding` is equally efficient(or less efficient but neglect-able) and preferred for simplicity and clarity in code.

Feel free to disagree and proceed with current approach. No strong opinion.